storage/stored_transaction.go

```diff
@@ -209,36 +227,44 @@ func (s *Storage) initBatchedTxsCounter() (err error) {
 	return err
 }
 
-func (s *TransactionStorage) GetTransactionHashesByBatchIDs(batchIDs ...models.Uint256) ([]common.Hash, error) {
-	hashes := make([]common.Hash, 0, len(batchIDs)*32)
+func (s *TransactionStorage) getTransactionHashesByBatchID(batchID models.Uint256) ([]common.Hash, error) {
+	hashes := make([]common.Hash, 0, 32)
 
-	// We have an index on CommitmentID. It turns out that BatchID is the first
-	// member of the BatchID struct, so we effectively have an index on BatchID.
-	// We can take advantage of that index by manually iterating.
-	// The slow version is: Badger.Find(..., bh.Where("CommitmentID.BatchID").In(...))
-
-	var keyList bh.KeyList
-	batchPrefixes := batchIdsToBatchPrefixes(batchIDs)
-	seekPrefix := db.IndexKeyPrefix(stored.BatchedTxName, "CommitmentID")
-	err := s.database.Badger.Iterator(seekPrefix, db.ReversePrefetchIteratorOpts, func(item *bdg.Item) (bool, error) {
-		if validForPrefixes(keyValue(seekPrefix, item.Key()), batchPrefixes) {
-			err := item.Value(func(val []byte) error {
-				return db.Decode(val, &keyList)
-			})
-			if err != nil {
-				return false, err
-			}
-			txHashes, err := decodeKeyListHashes(stored.BatchedTxPrefix, keyList)
-			if err != nil {
-				return false, err
-			}
-			hashes = append(hashes, txHashes...)
+	// nolint: gocritic
+	seekPrefix := append(stored.BatchedTxPrefix, batchID.Bytes()...)
+
+	// BatchedTx are stored with CommitmentSlot as their primary key: BatchID is the
+	// first member of CommitmentSlot which means we effectively have an index on
+	// BatchID.
+
+	var batchedTx stored.BatchedTx
+	err := s.database.Badger.Iterator(seekPrefix, db.PrefetchIteratorOpts, func(item *bdg.Item) (bool, error) {
+		err := item.Value(batchedTx.SetBytes)
+		if err != nil {
+			return false, err
 		}
+
+		hashes = append(hashes, batchedTx.Hash)
 		return false, nil
 	})
 	if err != nil && err != db.ErrIteratorFinished {
 		return nil, err
 	}
+
+	return hashes, nil
+}
+
+func (s *TransactionStorage) GetTransactionHashesByBatchIDs(batchIDs ...models.Uint256) ([]common.Hash, error) {
```

Comment:
`GetTransactionHashesByBatchIDs` and `MarkTransactionsAsPending` are only used together in `excludeTransactionsFromCommitment` in `revert_batches.go`. I think we can combine the two functions into one moving txs from `BatchedTx` to `PendingTx`. `MarkTransactionsAsPending` currently moves txs from `FailedTx` to `PendingTx` as well but I think we can drop this behaviour and make sure in other places that the same tx doesn't get written into two structures.

Reply:
Good point that they're only ever used together. I think it makes sense to keep these separated, each does one small thing that we might later want to do independently, and there's not really an efficiently gain from deleting as the same time as we iterate.

This PR does introduce some inefficiency though, just pushed a commit which fixes that by changing these methods to operate over slots instead of hashes.

Reply:
I suggested combining them to fix the inefficiency as well. After your refactor we can keep them separate 👍